File: adapters/introspect-mcp/src/korg_introspect_mcp/discovery.py
"""Discover a binary's callables by running `<binary> --introspect`
and parsing the `korg:introspect@v1` document.

Validation:
  - The document must declare `schema: "korg:introspect@v1"`.
  - It must include a `callables` array.
  - Each callable must carry `command_id`, `name`, `input_schema`, `capabilities`.

Discovery failures (binary not found, --introspect not supported, malformed
document) raise `DiscoveryError` with a clear message instead of partially-
populating the server. Better to fail loud at startup than serve broken tools.
"""

from __future__ import annotations

import json
import shutil
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Any


SUPPORTED_SCHEMA = "korg:introspect@v1"
DISCOVERY_TIMEOUT_S = 15.0


class DiscoveryError(RuntimeError):
    """Raised when the introspect document is missing, invalid, or unparseable."""


@dataclass
class DiscoveredCallable:
    """One callable, validated and ready to register as an MCP tool."""

    command_id: str
    name: str
    description: str
    input_schema: dict[str, Any]
    capabilities: dict[str, Any]
    surfaces: list[str]


@dataclass
class DiscoveredBinary:
    """The full discovery result for one binary."""

    binary_path: Path
    schema: str
    binary_name: str
    version: str
    callables: list[DiscoveredCallable]
    exit_codes: dict[str, str]
    callables_declared: bool

    def by_command_id(self, command_id: str) -> DiscoveredCallable | None:
        for c in self.callables:
            if c.command_id == command_id:
                return c
        return None


def resolve_binary(spec: str) -> Path:
    """Resolve a binary spec to an absolute path.

    Accepts: absolute path, relative path, or bare name (PATH lookup).
    """
    p = Path(spec).expanduser()
    if p.is_absolute():
        if not p.exists():
            raise DiscoveryError(f"binary not found at absolute path: {p}")
        return p
    # Relative or bare name → PATH lookup
    found = shutil.which(spec)
    if found is None:
        raise DiscoveryError(
            f"binary not found on PATH: {spec!r}. "
            f"If you meant a relative path, pass it as ./{spec}"
        )
    return Path(found)


def run_introspect(binary_path: Path, timeout_s: float = DISCOVERY_TIMEOUT_S) -> dict[str, Any]:
    """Exec `<binary> --introspect`, capture stdout, parse as JSON.

    stderr is captured but only surfaced on failure (matches the Foundry
    output contract: stdout = result, stderr = diagnostics).
    """
    try:
        result = subprocess.run(
            [str(binary_path), "--introspect"],
            capture_output=True,
            text=True,
            timeout=timeout_s,
            check=False,
        )
    except subprocess.TimeoutExpired as e:
        raise DiscoveryError(
            f"{binary_path} --introspect timed out after {timeout_s}s"
        ) from e
    except OSError as e:
        raise DiscoveryError(f"could not exec {binary_path}: {e}") from e

    if result.returncode != 0:
        # Some binaries print --introspect even on non-zero (e.g. clap's
        # missing-subcommand error). Try to parse anyway; only fail if
        # the parse also fails.
        try:
            return json.loads(result.stdout)
        except json.JSONDecodeError:
            raise DiscoveryError(
                f"{binary_path} --introspect exited {result.returncode}: "
                f"{result.stderr.strip() or result.stdout.strip()}"
            )

    try:
        return json.loads(result.stdout)
    except json.JSONDecodeError as e:
        raise DiscoveryError(
            f"{binary_path} --introspect did not return valid JSON: {e}. "
            f"First 200 chars: {result.stdout[:200]!r}"
        ) from e


def validate_document(doc: dict[str, Any], binary_path: Path) -> DiscoveredBinary:
    """Validate the document shape and return a typed DiscoveredBinary."""
    schema = doc.get("schema")
    if schema != SUPPORTED_SCHEMA:
        raise DiscoveryError(
            f"{binary_path}: unsupported introspect schema {schema!r}. "
            f"This version of korg-introspect-mcp supports {SUPPORTED_SCHEMA}."
        )

    raw_callables = doc.get("callables")
    if not isinstance(raw_callables, list):
        raise DiscoveryError(
            f"{binary_path}: introspect document has no 'callables' array."
        )

    discovered: list[DiscoveredCallable] = []
    seen_ids: set[str] = set()
    for i, c in enumerate(raw_callables):
        if not isinstance(c, dict):
            raise DiscoveryError(f"{binary_path}: callables[{i}] is not an object.")
        for required in ("command_id", "name", "input_schema", "capabilities"):
            if required not in c:
                raise DiscoveryError(
                    f"{binary_path}: callables[{i}] missing required field {required!r}"
                )
        cid = c["command_id"]
        if cid in seen_ids:
            raise DiscoveryError(f"{binary_path}: duplicate command_id {cid!r}")
        seen_ids.add(cid)

        discovered.append(
            DiscoveredCallable(
                command_id=cid,
                name=c["name"],
                description=c.get("description", ""),
                input_schema=c["input_schema"],
                capabilities=c["capabilities"],
                surfaces=list(c.get("surfaces") or []),
            )
        )

    return DiscoveredBinary(
        binary_path=binary_path,
        schema=schema,
        binary_name=doc.get("binary", binary_path.name),
        version=doc.get("version", "0.0.0"),
        callables=discovered,
        exit_codes=dict(doc.get("exit_codes") or {}),
        callables_declared=bool(doc.get("callables_declared", False)),
    )


def discover(spec: str, timeout_s: float = DISCOVERY_TIMEOUT_S) -> DiscoveredBinary:
    """Top-level: resolve binary spec, run --introspect, validate, return typed result."""
    binary_path = resolve_binary(spec)
    raw_doc = run_introspect(binary_path, timeout_s=timeout_s)
    return validate_document(raw_doc, binary_path)

File: adapters/introspect-mcp-ts/src/discovery.ts
// Discover a binary's callables by execing `<binary> --introspect`
// and validating the korg:introspect@v1 document.

import { spawn } from "node:child_process";
import { promises as fs } from "node:fs";
import { resolve as resolvePath } from "node:path";
import { homedir, platform } from "node:os";
import { delimiter } from "node:path";

export const SUPPORTED_SCHEMA = "korg:introspect@v1";
export const DISCOVERY_TIMEOUT_MS = 15_000;

export class DiscoveryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "DiscoveryError";
  }
}

export interface DiscoveredCallable {
  command_id: string;
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
  capabilities: Record<string, unknown>;
  surfaces: string[];
}

export interface DiscoveredBinary {
  binary_path: string;
  schema: string;
  binary_name: string;
  version: string;
  callables: DiscoveredCallable[];
  exit_codes: Record<string, string>;
  callables_declared: boolean;
}

export function findCallableByCommandId(
  binary: DiscoveredBinary,
  commandId: string
): DiscoveredCallable | null {
  return binary.callables.find((c) => c.command_id === commandId) ?? null;
}

// ── PATH lookup (Node 18 has no built-in shutil.which) ────────────────

function expanduser(p: string): string {
  if (p.startsWith("~")) return resolvePath(homedir() + p.slice(1));
  return resolvePath(p);
}

async function which(name: string): Promise<string | null> {
  // If it's a path (contains separator), test it directly.
  if (name.includes("/") || name.includes("\\")) {
    const p = expanduser(name);
    try {
      const stat = await fs.stat(p);
      if (stat.isFile()) return p;
    } catch {
      return null;
    }
    return null;
  }
  const pathEnv = process.env["PATH"] ?? "";
  const exts = platform() === "win32" ? (process.env["PATHEXT"] ?? "").split(";") : [""];
  for (const dir of pathEnv.split(delimiter)) {
    if (!dir) continue;
    for (const ext of exts) {
      const full = resolvePath(dir, name + ext);
      try {
        const stat = await fs.stat(full);
        if (stat.isFile()) return full;
      } catch {
        // try next
      }
    }
  }
  return null;
}

export async function resolveBinary(spec: string): Promise<string> {
  if (spec.includes("/") || spec.includes("\\") || spec.startsWith("~")) {
    const p = expanduser(spec);
    try {
      const stat = await fs.stat(p);
      if (stat.isFile()) return p;
    } catch {
      throw new DiscoveryError(`binary not found at path: ${p}`);
    }
    throw new DiscoveryError(`binary not a regular file: ${p}`);
  }
  const found = await which(spec);
  if (!found) {
    throw new DiscoveryError(
      `binary not found on PATH: ${JSON.stringify(spec)}. ` +
        `If you meant a relative path, pass it as ./${spec}`
    );
  }
  return found;
}

// ── Spawn + capture ───────────────────────────────────────────────────

interface SpawnResult {
  stdout: string;
  stderr: string;
  exitCode: number;
}

function execIntrospect(binaryPath: string, timeoutMs: number): Promise<SpawnResult> {
  return new Promise<SpawnResult>((resolve, reject) => {
    const child = spawn(binaryPath, ["--introspect"], {
      stdio: ["ignore", "pipe", "pipe"],
    });
    const stdout: Buffer[] = [];
    const stderr: Buffer[] = [];
    child.stdout.on("data", (b: Buffer) => stdout.push(b));
    child.stderr.on("data", (b: Buffer) => stderr.push(b));

    const timer = setTimeout(() => {
      child.kill("SIGTERM");
      reject(new DiscoveryError(`${binaryPath} --introspect timed out after ${timeoutMs}ms`));
    }, timeoutMs);

    child.on("error", (err) => {
      clearTimeout(timer);
      reject(new DiscoveryError(`could not exec ${binaryPath}: ${err.message}`));
    });
    child.on("close", (code) => {
      clearTimeout(timer);
      resolve({
        stdout: Buffer.concat(stdout).toString("utf8"),
        stderr: Buffer.concat(stderr).toString("utf8"),
        exitCode: code ?? -1,
      });
    });
  });
}

export async function runIntrospect(
  binaryPath: string,
  timeoutMs: number = DISCOVERY_TIMEOUT_MS
): Promise<unknown> {
  const result = await execIntrospect(binaryPath, timeoutMs);
  // Some clap-derived CLIs exit non-zero for --introspect (e.g. when other
  // required args are missing). Try to parse stdout regardless; only fail
  // if the parse also fails.
  try {
    return JSON.parse(result.stdout);
  } catch (e) {
    if (result.exitCode !== 0) {
      throw new DiscoveryError(
        `${binaryPath} --introspect exited ${result.exitCode}: ` +
          (result.stderr.trim() || result.stdout.trim() || "(no output)")
      );
    }
    throw new DiscoveryError(
      `${binaryPath} --introspect did not return valid JSON: ${(e as Error).message}. ` +
        `First 200 chars: ${JSON.stringify(result.stdout.slice(0, 200))}`
    );
  }
}

// ── Validation ────────────────────────────────────────────────────────

export function validateDocument(doc: unknown, binaryPath: string): DiscoveredBinary {
  if (typeof doc !== "object" || doc === null) {
    throw new DiscoveryError(`${binaryPath}: introspect document is not an object`);
  }
  const d = doc as Record<string, unknown>;
  const schema = d["schema"];
  if (schema !== SUPPORTED_SCHEMA) {
    throw new DiscoveryError(
      `${binaryPath}: unsupported introspect schema ${JSON.stringify(schema)}. ` +
        `Supported: ${SUPPORTED_SCHEMA}`
    );
  }
  const rawCallables = d["callables"];
  if (!Array.isArray(rawCallables)) {
    throw new DiscoveryError(`${binaryPath}: introspect document has no 'callables' array`);
  }

  const seenIds = new Set<string>();
  const callables: DiscoveredCallable[] = [];
  for (let i = 0; i < rawCallables.length; i++) {
    const c = rawCallables[i];
    if (typeof c !== "object" || c === null) {
      throw new DiscoveryError(`${binaryPath}: callables[${i}] is not an object`);
    }
    const co = c as Record<string, unknown>;
    for (const required of ["command_id", "name", "input_schema", "capabilities"]) {
      if (!(required in co)) {
        throw new DiscoveryError(
          `${binaryPath}: callables[${i}] missing required field ${JSON.stringify(required)}`
        );
      }
    }
    const cid = String(co["command_id"]);
    if (seenIds.has(cid)) {
      throw new DiscoveryError(`${binaryPath}: duplicate command_id ${JSON.stringify(cid)}`);
    }
    seenIds.add(cid);
    const surfacesRaw = co["surfaces"];
    callables.push({
      command_id: cid,
      name: String(co["name"]),
      description: typeof co["description"] === "string" ? co["description"] : "",
      input_schema: co["input_schema"] as Record<string, unknown>,
      capabilities: co["capabilities"] as Record<string, unknown>,
      surfaces: Array.isArray(surfacesRaw) ? surfacesRaw.map((s) => String(s)) : [],
    });
  }

  return {
    binary_path: binaryPath,
    schema,
    binary_name: String(d["binary"] ?? ""),
    version: String(d["version"] ?? "0.0.0"),
    callables,
    exit_codes: (d["exit_codes"] as Record<string, string>) ?? {},
    callables_declared: Boolean(d["callables_declared"]),
  };
}

export async function discover(
  spec: string,
  timeoutMs: number = DISCOVERY_TIMEOUT_MS
): Promise<DiscoveredBinary> {
  const binaryPath = await resolveBinary(spec);
  const rawDoc = await runIntrospect(binaryPath, timeoutMs);
  return validateDocument(rawDoc, binaryPath);
}
